File: gene_bank_service/internal/flora/service.go
//	Copyright 2025 Naveen R
//
//		Licensed under the Apache License, Version 2.0 (the "License");
//		you may not use this file except in compliance with the License.
//		You may obtain a copy of the License at
//
//		http://www.apache.org/licenses/LICENSE-2.0
//
//		Unless required by applicable law or agreed to in writing, software
//		distributed under the License is distributed on an "AS IS" BASIS,
//		WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//		See the License for the specific language governing permissions and
//		limitations under the License.

package flora

import (
	"fmt"
	"log"
	"project_chimera/gene_bank_service/internal/dto"
	"project_chimera/gene_bank_service/internal/rabbitmq"
	"project_chimera/gene_bank_service/pkg/utils"
	"project_chimera/gene_bank_service/pkg/utils/helpers"

	"github.com/gofiber/fiber/v2"
)

// FloraService defines the interface for flora services
type FloraService interface {
	GetFlora(c *fiber.Ctx) (dto.FloraResponse, error)
	GetFloraById(c *fiber.Ctx) (dto.FloraResponse, error)
	PostFlora(c *fiber.Ctx) error
	PutFlora(c *fiber.Ctx) error
	DeleteFlora(c *fiber.Ctx) error
}

// FloraService is the concrete implementation of FloraService
type floraService struct {
	upStreamHandler *rabbitmq.Handler

	downStreamHandler *rabbitmq.Handler

	errorQueueHandler *rabbitmq.Handler
}

func NewFloraService(upStreamHandler *rabbitmq.Handler, downStreamHandler *rabbitmq.Handler, errorQueueHandler *rabbitmq.Handler) FloraService {
	return &floraService{upStreamHandler: upStreamHandler, downStreamHandler: downStreamHandler, errorQueueHandler: errorQueueHandler}
}

// GetFlora handler for retrieving flora data
func (s *floraService) GetFlora(c *fiber.Ctx) (dto.FloraResponse, error) {
	res, err := s.downStreamHandler.SendRequest(c, "get_all_floras", "")
	if err != nil {
		log.Printf("Error in SendRequest: %v", err)

		data := map[string]interface{}{
			"code":   500,
			"status": "Internal Server Error",
			"type":   "GET",
			"data": map[string]interface{}{
				"error": err.Error(),
			},
		}

		s.errorQueueHandler.SendAckRequest(data, "flora.getall")
		return dto.FloraResponse{}, err
	}

	if res.Code != utils.SUCCESS {
		data := map[string]interface{}{
			"code":   res.Code,
			"status": res.Status,
			"type":   "GET",
			"data":   res.Data,
		}

		s.errorQueueHandler.SendAckRequest(data, "flora.getall")
		return dto.FloraResponse{}, helpers.HandleRPCError(res)
	}

	floraList, err := helpers.ProcessFloraData(res.Data)
	if err != nil {
		data := map[string]interface{}{
			"code":   500,
			"status": "Internal Server Error",
			"type":   "GET",
			"data": map[string]interface{}{
				"error": err.Error(),
			},
		}

		s.errorQueueHandler.SendAckRequest(data, "flora.getall")
		return dto.FloraResponse{}, err
	}

	return dto.FloraResponse{Flora: floraList}, nil
}

func (s *floraService) GetFloraById(c *fiber.Ctx) (dto.FloraResponse, error) {
	res, err := s.downStreamHandler.SendRequest(c, "get_flora_by_id", c.Params("id"))
	if err != nil {
		data := map[string]interface{}{
			"code":   500,
			"status": "Internal Server Error",
			"type":   "GET",
			"data": map[string]interface{}{
				"error": err.Error(),
			},
		}

		s.errorQueueHandler.SendAckRequest(data, "flora.getbyid")
		log.Printf("Error in SendRequest: %v", err)
		return dto.FloraResponse{}, err
	}

	if res.Code != utils.SUCCESS {
		data := map[string]interface{}{
			"code":   res.Code,
			"status": res.Status,
			"type":   "GET",
			"data":   res.Data,
		}

		s.errorQueueHandler.SendAckRequest(data, "flora.getbyid")
		return dto.FloraResponse{}, helpers.HandleRPCError(res)
	}

	floraList, err := helpers.ProcessFloraData(res.Data)
	if err != nil {
		data := map[string]interface{}{
			"code":   500,
			"status": "Internal Server Error",
			"type":   "GET",
			"data": map[string]interface{}{
				"error": err.Error(),
			},
		}

		s.errorQueueHandler.SendAckRequest(data, "flora.getbyid")
		return dto.FloraResponse{}, err
	}

	return dto.FloraResponse{Flora: floraList}, nil
}

// PostFlora handler for adding flora data
func (s *floraService) PostFlora(c *fiber.Ctx) error {
	var payload dto.FloraRequest

	if err := c.BodyParser(&payload); err != nil {
		data := map[string]interface{}{
			"code":   400,
			"status": "Bad Request",
			"type":   "POST",
			"data": map[string]interface{}{
				"error": err.Error(),
			},
		}

		s.errorQueueHandler.SendAckRequest(data, "flora.post")
		return &fiber.Error{Code: fiber.StatusBadRequest, Message: "Invalid JSON"}
	}

	// Handle image conversion to byte array
	var imageBytes []byte
	var err error

	if payload.ImageURL != "" {
		// If the image is provided via a URL, fetch it
		imageBytes, err = utils.FetchImageFromURL(payload.ImageURL)
	} else if payload.ImagePath != "" {
		// If the image is provided via a local path, read it
		imageBytes, err = utils.FetchImageFromPath(payload.ImagePath)
	} else if payload.Image != nil {
		// If the image is provided as a byte array, use it directly
		imageBytes = payload.Image
	} else {
		data := map[string]interface{}{
			"code":   400,
			"status": "Bad Request",
			"type":   "POST",
			"data": map[string]interface{}{
				"error": "No image provided",
			},
		}

		s.errorQueueHandler.SendAckRequest(data, "flora.post")
		// Handle case where there is no image provided
		return &fiber.Error{Code: fiber.StatusBadRequest, Message: "No image URL or path provided"}
	}

	if err != nil {
		data := map[string]interface{}{
			"code":   500,
			"status": "Internal Server Error",
			"type":   "POST",
			"data": map[string]interface{}{
				"error": err.Error(),
				"url":   payload.ImageURL,
			},
		}
		s.errorQueueHandler.SendAckRequest(data, "flora.post")
		return &fiber.Error{Code: fiber.StatusInternalServerError, Message: fmt.Sprintf("Error fetching image: %v", err)}
	}

	userId := c.Get("X-Auth-UserId")

	if userId == "" {
		data := map[string]interface{}{
			"code":   400,
			"status": "Bad Request",
			"type":   "POST",
			"data": map[string]interface{}{
				"error": "User ID not found in request",
			},
		}

		s.errorQueueHandler.SendAckRequest(data, "flora.post")
		return &fiber.Error{Code: fiber.StatusBadRequest, Message: "User ID not found in request"}
	}

	// Send Ack request
	var data map[string]interface{} = map[string]interface{}{
		"CommonName":     payload.CommonName,
		"ScientificName": payload.ScientificName,
		"Image":          imageBytes,
		"Description":    payload.Description,
		"Origin":         payload.Origin,
		"OtherDetails":   payload.OtherDetails,
		"type":           payload.Type,
		"UserId":         userId,
	}
	err = s.upStreamHandler.SendAckRequest(data, "add_flora")
	if err != nil {
		errData := map[string]interface{}{
			"code":   500,
			"status": "Internal Server Error",
			"type":   "POST",
			"data": map[string]interface{}{
				"error": err.Error(),
			},
		}

		s.errorQueueHandler.SendAckRequest(errData, "flora.post")
		return err
	}

	return nil
}

// PutFlora handler for updating flora data
func (s *floraService) PutFlora(c *fiber.Ctx) error {
	var payload dto.FloraUpdateRequest

	if err := c.BodyParser(&payload); err != nil {
		data := map[string]interface{}{
			"code":   400,
			"status": "Bad Request",
			"type":   "PUT",
			"data": map[string]interface{}{
				"error": err.Error(),
			},
		}

		s.errorQueueHandler.SendAckRequest(data, "flora.put")
		return &fiber.Error{Code: fiber.StatusBadRequest, Message: "Invalid JSON"}
	}

	// Handle image conversion to byte array
	var imageBytes []byte
	var err error

	if payload.ImageURL != "" {
		// If the image is provided via a URL, fetch it
		imageBytes, err = utils.FetchImageFromURL(payload.ImageURL)
	} else if payload.ImagePath != "" {
		// If the image is provided via a local path, read it
		imageBytes, err = utils.FetchImageFromPath(payload.ImagePath)
	} else if len(payload.Image) > 0 {
		imageBytes = payload.Image
	} else {
		// Handle case where there is no image provided
		data := map[string]interface{}{
			"code":   400,
			"status": "Bad Request",
			"type":   "PUT",
			"data": map[string]interface{}{
				"error": "No image provided",
			},
		}

		s.errorQueueHandler.SendAckRequest(data, "flora.put")
		return &fiber.Error{Code: fiber.StatusBadRequest, Message: "No image URL or path provided"}
	}

	if err != nil {
		data := map[string]interface{}{
			"code":   500,
			"status": "Internal Server Error",
			"type":   "PUT",
			"data": map[string]interface{}{
				"error": err.Error(),
				"url":   payload.ImageURL,
			},
		}

		s.errorQueueHandler.SendAckRequest(data, "flora.put")
		return &fiber.Error{Code: fiber.StatusInternalServerError, Message: fmt.Sprintf("Error fetching image: %v", err)}
	}

	userId := c.Get("X-Auth-UserId")

	if userId == "" {
		data := map[string]interface{}{
			"code":   400,
			"status": "Bad Request",
			"type":   "PUT",
			"data": map[string]interface{}{
				"error": "User ID not found in request",
			},
		}

		s.errorQueueHandler.SendAckRequest(data, "flora.put")
		return &fiber.Error{Code: fiber.StatusBadRequest, Message: "User ID not found in request"}
	}

	// Send Ack request
	data := utils.CreateFloraDataMap(payload, userId, imageBytes)
	err = s.upStreamHandler.SendAckRequest(data, "update_flora")
	if err != nil {
		errData := map[string]interface{}{
			"code":   500,
			"status": "Internal Server Error",
			"type":   "PUT",
			"data": map[string]interface{}{
				"error": err.Error(),
			},
		}

		s.errorQueueHandler.SendAckRequest(errData, "flora.put")
		return err
	}

	return nil
}

// DeleteFlora handler for deleting flora data
func (s *floraService) DeleteFlora(c *fiber.Ctx) error {
	return nil
}

File: flora_downstream_service/src/flora/service.py
# 	Copyright 2025 Naveen R
#
# 		Licensed under the Apache License, Version 2.0 (the "License");
# 		you may not use this file except in compliance with the License.
# 		You may obtain a copy of the License at
#
# 		http://www.apache.org/licenses/LICENSE-2.0
#
# 		Unless required by applicable law or agreed to in writing, software
# 		distributed under the License is distributed on an "AS IS" BASIS,
# 		WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# 		See the License for the specific language governing permissions and
# 		limitations under the License.

from typing import Any, Dict, Optional
import uuid
from sqlalchemy.future import select
from src.db.mongo.mongo_connect import mongo_engine
from src.db.postgres.postgres_connect import SessionLocal
from sqlalchemy.orm import Session
from fastapi import Depends
from sqlalchemy.ext.asyncio import AsyncSession
from src.model.common import FloraResponse
from src.model.flora import Flora
from src.model.mongo_flora import FloraMongo
from src.model.postgres_flora import FloraPG


async def get_floras(db: AsyncSession) -> Optional[FloraResponse]:
    try:
        result = await db.execute(select(FloraPG))
        florasPg: list[FloraPG] = result.scalars().all()

        if not florasPg:
            return FloraResponse(code=200, data=None)

        floras: list[Dict[str, Any]] = []
        for floraPg in florasPg:
            try:
                floraMongo = await mongo_engine.find_one(
                    FloraMongo, FloraMongo.flora_id == str(floraPg.id)
                )
            except Exception as e:
                print(f"An error occurred when retrieving Mongo data: {e}")
                return FloraResponse(
                    code=500,
                    data=f"An error occurred when retrieving Mongo data: {e}",
                )

            floraRes = Flora(
                id=str(floraPg.id),
                user_id=floraPg.user_id,
                common_name=floraPg.common_name,
                scientific_name=floraPg.scientific_name,
                type=floraPg.type,
                image=floraMongo.Image,
                description=floraMongo.Description,
                origin=floraMongo.Origin,
                other_details=floraMongo.OtherDetails,
            )
            floras.append(floraRes.model_dump_json())

        return FloraResponse(code=200, data=floras)

    except Exception as e:
        # Handle exceptions and log or raise them accordingly
        print(f"An error occurred: {e}")
        return FloraResponse(code=500, data=f"An error occurred: {e}")


async def get_flora(flora_id: str, db: AsyncSession) -> Optional[FloraResponse]:
    try:
        uid = uuid.UUID(flora_id)
    except ValueError:
        return FloraResponse(code=400, data="Invalid flora ID")

    result = await db.execute(select(FloraPG).where(FloraPG.id == uid))
    floraPg: FloraPG = result.scalar()
    if floraPg is None:
        return FloraResponse(code=404, data="Flora not found")

    floras: list[Dict[str, Any]] = []

    try:
        floraMongo = await mongo_engine.find_one(
            FloraMongo, FloraMongo.flora_id == str(flora_id)
        )
    except Exception as e:
        print(f"An error occurred when retrieving Mongo data: {e}")
        return FloraResponse(
            code=500,
            data=f"An error occurred when retrieving Mongo data: {e}",
        )

    floraRes = Flora(
        id=str(floraPg.id),
        user_id=floraPg.user_id,
        common_name=floraPg.common_name,
        scientific_name=floraPg.scientific_name,
        type=floraPg.type,
        image=floraMongo.Image,
        description=floraMongo.Description,
        origin=floraMongo.Origin,
        other_details=floraMongo.OtherDetails,
    )

    floras.append(floraRes.model_dump_json())

    return FloraResponse(code=200, data=floras)
